_posts/2016-08-05-lambda.markdown

```diff
@@ -0,0 +1,64 @@
+---
+layout: post
+title: Scaling with AWS Lambda
+date: 2016-08-05 20:35:04
+description: Scale your applications with AWS Lambda without having to think about it. This is a case study from a migration done at Lost My Name.
+---
+At [Lost My Name](https://www.lostmy.name/) I'm responsible for the reliability and performance of the platform. This platform is the core of the business as it enables the creatives to publish their work. This work is what makes Lost My Name's products so magical and popular.
+
+As of writing, the team is preparing for another Q4. This will be my second time going through this period at Lost My Name after been through it twice at [Boomf](https://boomf.com/). The experience is unique as you see the sales charts go up and up.
+
+Increased sales means increased prints we need to deliver to our printhouses. Every single one of hese prints goes though the platform to generate the personalised content at high resolution. Increased sales also means increased previews which also goes through the platform.
+
+In this blog post, I'll talk about a specific piece of one of our books, [The Incredible Intergalactic Journey Home](https://www.lostmy.name/personalized-products/the-journey-home-book).
+
+### The current situation
+
+For one of the spreads in the book - which exists out of 2 pages - we request your address. We then generate a page which consists of satellite imagery data to show kids their house and the area around their house. To do this, we use the [Bing Imagery API](https://msdn.microsoft.com/en-us/library/ff701721.aspx).
+
+To generate one page at print quality, we need to compose 4 Bing images at maximum resolution (2000x1500px). To have consistency - and some other non relevant reasons - between preview and print, we generate the same print quality image for the preview which we then resize before serving it to the user.
+
+This means that for every user that requests a book preview, we need to download 8 (1 spread, 2 pages) images, stitch them together and resize the image afterwards. All this work is quite CPU intensive and can't be cached (every request is unique due to it's specific coordinates).
+
+Illustrated below is the current state of the platform (simplified).
+
+![Imagery Data](/img/satellite-01.png)
+
+### The problem
+With the above architecture, we are able to serve roughly 2 books per second consistently without errors on [AWS EC2](https://aws.amazon.com/ec2/) t2.medium boxes. Crank up the concurrency and we see availability drop. High availability is crucial to us as we've seen that our preview drives conversion rates. 2 concurrent requests is not the traffic we want to serve. We want more, a lot more.
+
+As mentioned before, we're preparing for Q4 where our sales increase exponentially. To scale this, we would need a lot of boxes (we've done tests, changing to bigger boxes doesn't give us that much more performance).
```

Comment:
Unrelated vraag: als het effectief CPU-bound is, waarom geeft een box met meer CPU u dan niet meer performance?
Ik las eerder dat ge t2.medium boxes gebruikt. Waarom geen c4 (compute-optimized)?
Hebt ge t2.medium -> t2.large geprobeerd als bigger boxes? Want beide hebben maar zelfde hoeveelheid CPU (memory is voornaamste verschil), dus in dit geval zou je inderdaad geen hogere throughput kunnen verwachten.